enarx-keep/src/main.rs

```diff
@@ -0,0 +1,102 @@
+// SPDX-License-Identifier: Apache-2.0
+
+//! Loads keeps from their backends.
+
+#![deny(clippy::all)]
+#![deny(missing_docs)]
+
+macro_rules! debug {
+    ($($token:tt)+) => {
+        #[cfg(debug_assertions)]
+        eprintln!($($token)+)
+    };
+}
+
+mod backend;
+mod binary;
+
+use std::io::{Error, ErrorKind, Result};
+
+use backend::{Backend, Command};
+use binary::Component;
+
+fn info(backends: &[Box<dyn Backend>], _: std::env::Args) -> Result<()> {
+    use colorful::*;
+
+    for backend in backends {
+        println!("Backend: {}", backend.name());
+
+        let data = backend.data();
+
+        for datum in &data {
+            let icon = match datum.pass {
+                true => "✔".green(),
+                false => "✗".red(),
+            };
+
+            if let Some(info) = datum.info.as_ref() {
+                println!(" {} {}: {}", icon, datum.name, info);
+            } else {
+                println!(" {} {}", icon, datum.name);
+            }
+        }
+
+        for datum in &data {
+            if let Some(mesg) = datum.mesg.as_ref() {
+                println!("\n{}\n", mesg);
+            }
+        }
+    }
+
+    Ok(())
+}
+
+fn exec(backends: &[Box<dyn Backend>], code: String) -> Result<()> {
+    let code = Component::from_path(code)?;
+
+    for backend in backends {
+        if !backend.have() {
+            debug!("Backend {} not supported. Skipping...", backend.name());
+            continue;
+        }
+
+        let shim = Component::from_path(&backend.shim()?)?;
+        let keep = backend.build(shim, code)?;
+
+        let mut thread = keep.lock().unwrap().add_thread()?;
+        loop {
+            match thread.enter()? {
+                Command::SysCall(block) => unsafe {
+                    block.msg.rep = block.msg.req.syscall();
+                },
+            }
+        }
+    }
+
+    Err(Error::new(ErrorKind::NotFound, "No backend found!"))
+}
+
+fn main() -> Result<()> {
+    let backends: Vec<Box<dyn Backend>> = vec![Box::new(backend::sgx::Backend)];
+
+    let mut args = std::env::args();
+
+    if let Some(arg) = args.nth(1) {
+        match arg.as_str() {
+            "exec" => {
+                if let Some(code) = args.next() {
+                    return exec(&backends, code);
+                }
+            }
+
+            "info" => return info(&backends, args),
+
+            _ => (),
+        }
+    }
+
+    eprintln!("Usage:");
+    eprintln!("    enarx-keep info");
+    eprintln!("    enarx-keep exec <code>");
```

Comment:
maybe sometimes you want to specify the keep manually ? like the debug kvm keep without encryption?

Reply:
Added.